wagtaildraftail/widgets.py

```diff
@@ -22,11 +30,47 @@ class DraftailTextArea(WidgetWithScript, forms.HiddenInput):
     """
     Field widget to render a rich text editor powered by Draftail.
     """
-    def __init__(self, attrs=None, options=None):
-        self.options = self.intercept_image_formats(options or {})
+    # this class's constructor accepts a 'features' kwarg
+    accepts_features = True
+
+    def __init__(self, attrs=None, options=None, features=None):
+        # Find or construct an 'options' dict for this editor to use, according to
+        # this order of precedence:
+        # 1) If we receive an explicit 'features' list (and we're on a Wagtail version
+        #    that supports it), build options from that
+        # 2) If we receive an 'options' dict that looks like it's configuring things
+        #    longhand (i.e. contains any of 'entityTypes' / 'blockTypes' / 'inlineStyles'),
+        #    use that
+        # 3) If we're on a Wagtail version that supports rich text features,
+        #    build options from the default feature set
+        # 4) Otherwise, use whatever 'options' dict we have
```

Comment:
I'm hoping that introducing features will eliminate the need for a low-level `blockTypes`-style configuration in `options` - if so, this would only be an issue for backwards compatibility.

Even then, I can't really see any situations where this order of precedence between `options` and `features` is going to be confusing or break user expectations. For example, it would be reasonable enough for a user to configure the `default_draftail` editor the way that they want using `blockTypes` etc, and use plain `DraftailTextField()` most of the time, but occasionally use `DraftailTextField(features=[...])` when they want to depart from the standard setup. In this case it's clear that `features` is overriding the `options` config - we wouldn't really want to output a warning for that.

The one place the behaviour _might_ be surprising is if the OPTIONS config contains both a `features` list and a `blockTypes`-style configuration... but I'm not sure why anyone would do that, or why one alternative would be considered more correct than the other. (However, it is the one case where writing a system check would be relatively easy...)

Reply:
Fair point.